File: claude/.claude/hooks/pr-refresh-reminder.py
#!/usr/bin/env python3
"""PostToolUse hook (Bash): after a `git push`, if the branch has an open PR,
remind to refresh its title/body via the `pr-describer` agent.

Global rule: a push changes the diff, so an existing PR's description is stale.
This is the "partial" bucket — "a PR exists" is checkable (gh), "the description
is now stale" is the inference we nudge on. Reminder only, never blocks.

FAIL-OPEN: any error (no gh, no PR, not a repo) → no output.
"""
import sys
import json
import os
import re
import subprocess


def emit_context(text):
    json.dump({"hookSpecificOutput": {
        "hookEventName": "PostToolUse",
        "additionalContext": text,
    }}, sys.stdout)


def main():
    data = json.load(sys.stdin)
    if data.get("tool_name") != "Bash":
        return
    cmd = (data.get("tool_input") or {}).get("command")
    if not isinstance(cmd, str) or not re.search(r"\bgit\s+push\b", cmd):
        return
    cwd = data.get("cwd") or os.getcwd()
    r = subprocess.run(
        ["gh", "pr", "view", "--json", "url,number"],
        capture_output=True, text=True, timeout=15, cwd=cwd,
    )
    if r.returncode != 0:
        return  # no PR for this branch, or gh unavailable
    try:
        pr = json.loads(r.stdout)
        url = pr.get("url")
    except Exception:
        return
    if not url:
        return
    emit_context(
        f"Push landed and this branch has an open PR ({url}). Its diff changed, "
        f"so the description is now stale — delegate a refresh of the title/body "
        f"to the `pr-describer` agent (never edit the description inline)."
    )


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass  # fail-open

File: opencode/.config/opencode/plugins/pr-refresh-reminder.ts
import {
    extractOpencodeAfter,
    applyOpencodeAfter,
    type HookResult,
    type HookCtx,
    type HookInput,
} from "../hook-lib/hook-utils"

import { execFile } from "node:child_process"
import { promisify } from "node:util"

const exec = promisify(execFile)

// After a `git push`, if the branch has an open PR, remind to refresh its stale
// title/body via the pr-describer agent. Reminder only, never blocks. FAIL-OPEN:
// any error (no gh, no PR, not a repo) → none.

const PUSH = /\bgit\s+push\b/

function isGitPush(cmd: string): boolean {
    return PUSH.test(cmd)
}

function refreshContext(url: string): string {
    return (
        `Push landed and this branch has an open PR (${url}). Its diff changed, ` +
        `so the description is now stale — delegate a refresh of the title/body ` +
        `to the \`pr-describer\` agent (never edit the description inline).`
    )
}

async function run(input: HookInput, ctx: HookCtx): Promise<HookResult> {
    const cmd = input.command
    if (typeof cmd !== "string" || !isGitPush(cmd)) return { kind: "none" }
    try {
        const { stdout } = await exec("gh", ["pr", "view", "--json", "url,number"], {
            cwd: ctx.directory,
            timeout: 15000,
        })
        const url = JSON.parse(stdout).url
        return url ? { kind: "context", text: refreshContext(url) } : { kind: "none" }
    } catch {
        return { kind: "none" }
    }
}

export const prRefreshReminder = async ({ directory }: { directory: string }) => ({
    "tool.execute.after": async (input: any, output: any) =>
        applyOpencodeAfter(output, await run(extractOpencodeAfter(input, output), { directory })),
})
